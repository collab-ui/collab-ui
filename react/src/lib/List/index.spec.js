import React from 'react';
import { shallow, mount } from 'enzyme';
import {
  List,
  ListItem,
  SelectOption,
  SpaceListItem,
} from '@collab-ui/react';

describe('tests for <List />', () => {
  it('should match SnapShot', () => {
    const container = shallow(<List id="test" />);

    expect(container).toMatchSnapshot();
  });

  it('should render one List', () => {
    const container = mount(<List />);

    expect(container.find('.cui-list').length).toEqual(1);
  });

  it('should handle className prop', () => {
    const container = mount(<List className='menuItem'/>);

    expect(container.find('.cui-list').hasClass('menuItem')).toEqual(true);
  });

  it('should handle role prop', () => {
    const container = shallow(<List role='menuitem'/>);

    expect(container.find('.cui-list').props().role).toEqual('menuitem');
  });

  it('should handle tabType prop', () => {
    const container = shallow(
      <List tabType='horizontal'>
        <ListItem />
      </List>
    );

    expect(container.find('.cui-list--horizontal').length).toEqual(1);
  });

  it('should handle wrap prop', () => {
    const container = shallow(
      <List tabType='horizontal' wrap>
        <ListItem />
      </List>
    );

    expect(container.find('.cui-list--wrap').length).toEqual(1);
  });

  it('should handle itemRole prop', () => {
    const container = mount(
      <List itemRole='newRole'>
        <ListItem />
      </List>
    );
    
    expect(container.find('[role="newRole"]').length).toEqual(1);
  });

  it('should handle type prop', () => {
    const container = mount(
      <List type='small'>
        <ListItem />
      </List>
    );

    expect(container.find('.cui-list-item--small').length).toEqual(1);
  });

  it('should render children', () => {
    const container = shallow(
      <List>
        <div className="testingforTC" />
      </List>
    );

    expect(container.find('.testingforTC').length).toEqual(1);
  });

  it('should handle onSelect event', () => {
    const onSelect = jest.fn();

    const container = mount(
      <List onSelect={onSelect}>
        <ListItem label="test" link='javscript:void(0)' />
        <ListItem className='secondIndex' link='javscript:void(0)' />
      </List>
    );

    container.find('.secondIndex').first().simulate('click');
    expect(onSelect).toHaveBeenCalled();
  });

  it('should handle keyPress event (Up/Down/Left/Right)', () => {
    const container = mount(
      <List>
        <ListItem className='firstIndex' label="test" id='test-list-1' link='javscript:void(0)' />
        <ListItem className='secondIndex' label="test" id='test-list-2' link='javscript:void(0)' />
        <ListItem className='thirdIndex' label="test" id='test-list-3' link='javscript:void(0)' />
      </List>
    );
    container.update();
    container.update();

    const anchor1 = container.find('.firstIndex').first();

    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor1.simulate('keydown', {keyCode: 40, which: 40, charCode: 40});
    expect(container.state().listContext.focus).toEqual('test-list-2');
    anchor1.simulate('keydown', {keyCode: 39, which: 39, charCode: 39});
    expect(container.state().listContext.focus).toEqual('test-list-2');

    anchor1.simulate('keydown', {keyCode: 38, which: 38, charCode: 38});
    expect(container.state().listContext.focus).toEqual('test-list-3');
    anchor1.simulate('keydown', {keyCode: 37, which: 37, charCode: 37});
    expect(container.state().listContext.focus).toEqual('test-list-3');
  });

  it('should handle keyPress event (PageUp/PageDown/Home/End)', () => {
    const container = mount(
      <List>
        <ListItem className='firstIndex' label="test" id='test-list-1' link='javscript:void(0)' />
        <ListItem className='secondIndex' label="test" id='test-list-2' link='javscript:void(0)' />
        <ListItem className='thirdIndex' label="test" id='test-list-3' link='javscript:void(0)' />
      </List>
    );

    const anchor1 = container.find('.firstIndex').first();
    const anchor3 = container.find('.thirdIndex').first();

    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor1.simulate('keydown', {keyCode: 34, which: 34, charCode: 34});
    expect(container.state().listContext.focus).toEqual('test-list-3');
    anchor1.simulate('keydown', {keyCode: 35, which: 35, charCode: 35});
    expect(container.state().listContext.focus).toEqual('test-list-3');

    anchor3.simulate('keydown', {keyCode: 33, which: 33, charCode: 33});
    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor3.simulate('keydown', {keyCode: 36, which: 36, charCode: 36});
    expect(container.state().listContext.focus).toEqual('test-list-1');
  });

  it('should handle keyPress event (o)', () => {
    const container = mount(
      <List>
        <ListItem className='firstIndex' label="arrest" link='javscript:void(0)' id='test-list-1' />
        <ListItem className='secondIndex' label="detest" link='javscript:void(0)' id='test-list-2'/>
        <ListItem className='thirdIndex' label="obsessed" link='javscript:void(0)' id='test-list-3' />
      </List>
    );

    const anchor1 = container.find('.firstIndex').first();

    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor1.simulate('keydown', {keyCode: 79, which: 79, charCode: 79, key: 'o'});
    expect(container.state().listContext.focus).toEqual('test-list-3');
  });

  it('should loop by default', () => {
    const container = mount(
      <List>
        <ListItem className='firstIndex' label="test" id='test-list-1' link='javscript:void(0)' />
        <ListItem className='secondIndex' label="test" id='test-list-2' link='javscript:void(0)' />
        <ListItem className='thirdIndex' label="test" id='test-list-3' link='javscript:void(0)' />
      </List>
    );

    const anchor1 = container.find('.firstIndex').first();

    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor1.simulate('keydown', {keyCode: 38, which: 38, charCode: 38});
    expect(container.state().listContext.focus).toEqual('test-list-3');
  });

  it('should not loop when shouldLoop={false}', () => {
    const container = mount(
      <List shouldLoop={false}>
        <ListItem className='firstIndex' label="test" id='test-list-1' link='javscript:void(0)' />
        <ListItem className='secondIndex' label="test" id='test-list-2' link='javscript:void(0)' />
        <ListItem className='thirdIndex' label="test" id='test-list-3' link='javscript:void(0)' />
      </List>
    );

    const anchor1 = container.find('.firstIndex').first();

    expect(container.state().listContext.focus).toEqual('test-list-1');
    anchor1.simulate('keydown', {keyCode: 38, which: 38, charCode: 38});
    expect(container.state().listContext.focus).toEqual('test-list-1');
  });

  it('should handle SpaceListItem child', () => {
    const container = mount(
      <List>
        <SpaceListItem link='javscript:void(0)' header='header' id='test-list-1'/>
      </List>
    );

    expect(container.find('.cui-list-item--space').exists()).toEqual(true);
    expect(container.state().listContext.focus).toEqual('test-list-1');
    expect(container.state().listContext.active).toEqual(null);
  });

  it('should overwrite state with active prop', () => {
    const container = mount(
      <List active={'test-list-2'}>
        <SpaceListItem link='javscript:void(0)' header='header' id='test-list-1'/>
        <SelectOption link='javscript:void(0)' id='test-list-2'/>
      </List>
    );

    expect(container.find('.cui-list-item').at(0).hasClass('active')).toEqual(false);
    expect(container.find('.cui-list-item').at(1).hasClass('active')).toEqual(true);
  });

  it('should handle navigation on disabled children', () => {
    const container = mount(
      <List>
        <SpaceListItem className='firstIndex' link='javscript:void(0)' header='header' disabled/>
        <SpaceListItem className='secondIndex' link='javscript:void(0)' header='header' id='test-list-1'/>
        <SpaceListItem className='thirdIndex' link='javscript:void(0)' header='header' disabled/>
        <SpaceListItem className='fourthIndex' link='javscript:void(0)' header='header' id='test-list-2'/>
        <SpaceListItem className='fifthIndex' link='javscript:void(0)' header='header' disabled/>
      </List>
    );

    expect(container.state().listContext.focus).toEqual('test-list-1');
    const anchor2 = container.find('.secondIndex').first();
    anchor2.simulate('keydown', {keyCode: 40, which: 40, charCode: 40});
    const anchor4 = container.find('.fourthIndex').first();
    expect(container.state().listContext.focus).toEqual('test-list-2');
    anchor4.simulate('keydown', {keyCode: 40, which: 40, charCode: 40});
    expect(container.state().listContext.focus).toEqual('test-list-1');
  });

  it('should handle navigation on readOnly children', () => {
    const container = mount(
      <List>
        <SpaceListItem className='firstIndex' link='javscript:void(0)' header='header' isReadOnly/>
        <SpaceListItem className='secondIndex' link='javscript:void(0)' header='header' id='test-list-1'/>
        <SpaceListItem className='thirdIndex' link='javscript:void(0)' header='header' isReadOnly/>
        <SpaceListItem className='fourthIndex' link='javscript:void(0)' header='header' id='test-list-2'/>
        <SpaceListItem className='fifthIndex' link='javscript:void(0)' header='header' isReadOnly/>
      </List>
    );

    expect(container.state().listContext.focus).toEqual('test-list-1');
    const anchor2 = container.find('.secondIndex').first();
    anchor2.simulate('keydown', {keyCode: 40, which: 40, charCode: 40});
    expect(container.state().listContext.focus).toEqual('test-list-2');
    const anchor4 = container.find('.fourthIndex').first();
    anchor4.simulate('keydown', {keyCode: 40, which: 40, charCode: 40});
    expect(container.state().listContext.focus).toEqual('test-list-1');
  });

  it('should handle focusFirst prop', () => {
    const container = mount(
      <List focusFirst={false}>
        <ListItem className='firstIndex' label="test" link='javscript:void(0)' />
        <ListItem className='secondIndex' label="test" link='javscript:void(0)' />
        <ListItem className='thirdIndex' label="test" link='javscript:void(0)' />
      </List>
    );
    container.update();

    expect(container.state().listContext.focus).toEqual(null);
  });
});
