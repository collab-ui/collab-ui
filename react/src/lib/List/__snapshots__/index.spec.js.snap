// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <List /> should match SnapShot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List
    active={null}
    className=""
    focusFirst={true}
    id="test"
    itemRole="listitem"
    onSelect={null}
    role="list"
    shouldFocusActive={false}
    shouldLoop={true}
    tabType="vertical"
    trackActive={true}
    type={null}
    wrap={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Context.Provider
        value={
          Object {
            "active": null,
            "focus": null,
            "role": "listitem",
            "type": null,
          }
        }
      >
        <div
          aria-activedescendant={null}
          className="cui-list cui-list--vertical"
          id="test"
          onSelect={null}
          role="list"
        >
          <Unknown />
        </div>
      </Context.Provider>,
      "value": Object {
        "parentKeyDown": [Function],
        "parentOnSelect": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          aria-activedescendant={null}
          className="cui-list cui-list--vertical"
          id="test"
          onSelect={null}
          role="list"
        >
          <Unknown />
        </div>,
        "value": Object {
          "active": null,
          "focus": null,
          "role": "listitem",
          "type": null,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-activedescendant": null,
          "children": <Unknown />,
          "className": "cui-list cui-list--vertical",
          "id": "test",
          "onSelect": null,
          "role": "list",
        },
        "ref": [Function],
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Context.Provider
          value={
            Object {
              "active": null,
              "focus": null,
              "role": "listitem",
              "type": null,
            }
          }
        >
          <div
            aria-activedescendant={null}
            className="cui-list cui-list--vertical"
            id="test"
            onSelect={null}
            role="list"
          >
            <Unknown />
          </div>
        </Context.Provider>,
        "value": Object {
          "parentKeyDown": [Function],
          "parentOnSelect": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div
            aria-activedescendant={null}
            className="cui-list cui-list--vertical"
            id="test"
            onSelect={null}
            role="list"
          >
            <Unknown />
          </div>,
          "value": Object {
            "active": null,
            "focus": null,
            "role": "listitem",
            "type": null,
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "aria-activedescendant": null,
            "children": <Unknown />,
            "className": "cui-list cui-list--vertical",
            "id": "test",
            "onSelect": null,
            "role": "list",
          },
          "ref": [Function],
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": Object {
          "$$typeof": Symbol(react.provider),
          "_context": Object {
            "$$typeof": Symbol(react.context),
            "Consumer": Object {
              "$$typeof": Symbol(react.context),
              "_calculateChangedBits": null,
              "_context": [Circular],
            },
            "Provider": [Circular],
            "_calculateChangedBits": null,
            "_currentRenderer": null,
            "_currentRenderer2": null,
            "_currentValue": null,
            "_currentValue2": null,
            "_threadCount": 0,
          },
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": undefined,
          "_currentValue2": undefined,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
